src/Symfony/Component/HttpKernel/CacheWarmer/CacheWarmerAggregate.php

```diff
@@ -96,7 +96,7 @@ public function warmUp(string $cacheDir, SymfonyStyle $io = null): array
                 }
 
                 $start = microtime(true);
-                foreach ((array) $warmer->warmUp($cacheDir) as $item) {
+                foreach ((array) $warmer->warmUp($cacheDir, null === $buildDir || $warmer->isOptional() ? null : $buildDir) as $item) {
```

Comment:
What's the rationale for tightening optional+build-dir concepts together? I'd prefer removing this extra check, unless reason.

Reply:
The check in itself is normally useless, because non-optional and optional warmers are never run together from the same place.
Non-optional warmers are always run at the end of `Symfony\Component\HttpKernel\Kernel::initializeContainer`.
Optional warmers (and only them) are run from the `CacheClearCommand` and `CacheWarmupCommand`.

The rationale of coupling optional to build-dir concepts is more about coupling the build-dir to the moment you run the warmers and the potential write state of the build-dir. It seemed more natural to me to consider the build-dir only writeable when the Kernel files are being compiled. The easiest way to know that the Kernel is compiling is to restrict the build-dir to non-optional warmers. Do you have any objection to that rationale ?

Writing this, I realise we could also create a third state with optional warmer with build-dir access, that would be enabled only when running `CacheClearCommand`, but I'm wondering if it is possible to end up in a situation where `CacheClearCommand` is executed with a read-only build-dir ?

Reply:
After some reflexion time on my side, and playing a bit with possibilities, I removed completely the coupling between the optional and build-dir concepts, as you were suggesting 👍